### PR TITLE
add constants api to codegen

### LINKED ISSUE #396: Generate API for pallet constants
https://github.com/paritytech/subxt/pull/207 introduced a getter for `PalletMetadata` but as the strings are not validated at compile-time they may eventually fall out-of-sync. It would be better it add a `ConstantsApi` in the codegen crate.

### PR DESCRIPTION
**Pull request by gregdhill**
Signed-off-by: Gregory Hill <gregorydhill@outlook.com>

Closes #396

### CODE REVIEW

**Comment by gregdhill** [comment]
@ascjones @jsdw what do you think of this approach? Do you have any recommendations?

**Comment by jsdw** [comment]
Offhand I'm happy with this approach; it looks fairly clean and I like the api that you end up with!

File: codegen/src/api/constants.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                Ok(::subxt::codec::Decode::decode(&mut &vec![#(#ref_slice,)*][..])?)
+            }
+        }
+    });
+
+    quote! {
+        pub mod constants {
+            use super::#types_mod_ident;
+
+            pub struct ConstantsApi { }
```
Comment (ascjones) [inline codegen/src/api/constants.rs:49]:
```suggestion
            pub struct ConstantsApi;
```

File: codegen/src/api/constants.rs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+            impl ConstantsApi {
+                pub fn new() -> Self {
+                    Self { }
+                }
```
Comment (ascjones) [inline codegen/src/api/constants.rs:54]: Don't really need this constructor

**Review by ascjones — APPROVED** [review]
LGTM!

File: codegen/src/api/constants.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+use proc_macro2::TokenStream as TokenStream2;
+use quote::{
+    format_ident,
+    quote,
+};
+use scale_info::form::PortableForm;
+
+pub fn generate_constants(
+    type_gen: &TypeGenerator,
+    constants: &Vec<PalletConstantMetadata<PortableForm>>,
```
Comment (ascjones) [inline codegen/src/api/constants.rs:29]:
```suggestion
    constants: &[PalletConstantMetadata<PortableForm>],
```
To satisfy clippy

File: codegen/src/api/constants.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+) -> TokenStream2 {
+    let constant_fns = constants.iter().map(|constant| {
+        let fn_name = format_ident!("{}", constant.name.to_snake_case());
+        let return_ty = type_gen.resolve_type_path(constant.ty.id(), &[]);
+
+        let ref_slice = constant.value.as_slice();
+
+        quote! {
+            pub fn #fn_name(&self) -> ::core::result::Result<#return_ty, ::subxt::BasicError> {
+                Ok(::subxt::codec::Decode::decode(&mut &vec![#(#ref_slice,)*][..])?)
```
Comment (jsdw) [inline codegen/src/api/constants.rs:40]: Would this work to save an allocation?

```suggestion
                Ok(::subxt::codec::Decode::decode(&mut &[#(#ref_slice,)*])?)
```

**Review by jsdw — APPROVED** [review]

**Pull request merged**